src/context.ts

```diff
@@ -117,4 +125,29 @@ class Context {
   }
 }
 
-module.exports = Context
+export interface PayloadRepository {
```

Comment:
this interface is missing a lot of attributes that exist on some payloads, which makes typescript complains about non existent properties :(
eg: `id` is almost always available, so is `html_url`, ...
this makes this type unusable in most of the cases, and I have to override it which is super annoying
all the properties should be available here, and those that are not always present should be conditional (with a `?`).

Reply:
Another solution is to add `[key: string]: any;` to each property so that it can include any other property and we don't have to define all of them, but only the ones defined here will be available for auto complete.
Or maybe use a generic type `<T>`?

Reply:
Yeah, I think we're going to need types for GitHub webhook payloads.

Reply:
@ocombe I was going through some possibilities here and wanted to get your thoughts. The problem I see with trying to map types for the webhook payloads is that the schema is not consistent. Even for a given webhook `event`, the `action` might change the schema.

Example:

### IssuesEvent
Triggered when an issue is assigned, unassigned, labeled, unlabeled, opened, edited, milestoned, demilestoned, closed, or reopened.

key | type | description
-- | -- | --
`changes` | `object` | The changes to the issue **if the action was "edited"**.

So if you have:

`robot.on('issues.edited', context => {..}`

`context.payload` will have the `changes` object, but otherwise it will not.

Would it be acceptable to just stop at defining `context.payload` as any Object (`any{}`)? I don't know that even with the new [Conditional Types](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html), we can define all the potential payload types without the type definition being overly complex.

I'm definitely not against trying to do that, I just don't know if the effort is necessary. Thoughts?

Reply:
Ah, I like the first one! I'll go through the webhook event types to see how to define the  "generic" form, but then I think it would make the most sense to have a type for each Event, with optional types inside it:

`robot.on('issues.created', (context: IssuesContext) => {...}`
`robot.on('issue_comment.created', (context: IssueCommentContext) => {...}`

Then each Event Type Context would have a `changes?` key.